proposals/0016-simd-program-rebatable-account-write-fees.md

```diff
@@ -0,0 +1,621 @@
+through runtime features. In particular, programs needs to be able to penalize,
+even if the application is not CPI'd to.
+
+The checks on the PRAW fees will be taken care of by Solana runtime. Total PRAW
+fees paid will be included in the transaction through a special
+`SetProgramRebatableAccountWriteFees` similar to the `ComputeBudget` program so
+that it is easier to calculate the total amount of fees that will be paid. In
+case of nonpayment or partial payment, the program is never executed. This
+instruction cannot be CPI'ed into.
+
+Addition of new syscalls to get, set and rebate PRAW fees during program
+runtime. Application developers will be able to rebate these fees through the
+new syscall, but only if the transaction succeeds, this way transactions
+write-locking the respective accounts will require full fee payment before
+program is executed.
+
+Once PRAW fees are paid on an account by the payer they are valid for the whole
+transaction, even if the same account is write-locked in multiple instructions.
+
+### Payment
+
+`SetProgramRebatableAccountWriteFees` will be part of the updated compute budget
+program. `SetProgramRebatableAccountWriteFees` will take a u64 parameter to set
+microlamports per requested CU as the PRAW fees. Instead of having separate
+instructions for each compute budget parameter and PRAW fee we will have a
+single instruction that will set all the compute budget and PRAW fee parameters.
+Compute budget parameters could be considered as transaction header that sets
+some parameters for the transaction. We create a new solana native program
+`TransactionHeaderProgram` that decodes the instruction data using a type length
+value structure (`TLV`) format and computes the required
+`TransactionHeaderParameter`s for the transaction. The `TLV` structure is well
+defined for accounts in solana program library repository we will extend the
+definition for instruction data. As we have multiple transaction header
+parameters we convert instruction data into a list of
+`TransactionHeaderParameter`. In case we fail to deserialize the transaction
+data, the transaction will fail with an error `InvalidInstructionData`. If same
+`TransactionHeaderParameter` is used multiple times or has multiple instructions
+using `TransactionHeaderProgram` we return an error `DuplicateInstruction`.
+Similarly, if we use old Compute Budget instructions with the new instruction to
+set the same parameters we return an error `DuplicateInstruction`.
+
+Program Id for the program `TransactionHeaderProgram`:
+
+```
+TrasactionHeader111111111111111111111111111
+```
+
+Definition for transaction header parameters:
+
+```
+enum TransactionHeaderParameter {
+  // existing compute budget parameters
+  RequestHeapFrame(u32),
+  SetComputeUnitLimit(u32),
+  SetComputeUnitPrice(u64),
+  SetLoadedAccountsDataSizeLimit(u32),
+  // Limit on maximum of PRAW fees to be paid
+  // Fees is set in micro lamports per CU
```

Comment:
Maybe you meant to use microlamports to be in consistency with ComputeUnitPrice? Otherwise, introducing `TransactionHeaderParameter` might be a good opportunity to put term `microlamports` in rest. @mvines @t-nelson wdyt?

Reply:
Updated to microlamports to be consistent.